File: mobile/src/api/placeSearch.ts
import { apiRequest } from './client';

/** One real, named place a guide might mean — from
 * backend/app/services/places/ (Google Places). `label` is the full display
 * string; the app never assembles its own from raw parts. */
export interface PlaceSearchResult {
  label: string;
  latitude: number;
  longitude: number;
  /** The provider's own id for this exact place, when available. Carried
   * through as externalPlaceId if the guide picks this result — see
   * MemoryContributeScreen.handleSelectPlace. */
  placeId: string | null;
}

interface PlaceSearchResponseWire {
  results: { label: string; latitude: number; longitude: number; place_id: string | null }[];
}

/**
 * GET /api/v1/guides/{guideId}/place-search?q=... — place-name autocomplete
 * for describing a memory/old photo without exact coordinates. Read-only,
 * stateless: selecting a result is an on-device concern (see
 * src/components/PlaceAutocomplete.tsx). Biased toward this guide's own
 * recorded location history server-side, never restricted to it.
 *
 * Throws NetworkError/ApiError like every other endpoint module — callers
 * treat "search failed" as "no suggestions right now", never as a reason to
 * block saving a memory (see the offline-first rule: place search is a
 * convenience, not a requirement).
 */
export async function searchPlaces(guideId: string, query: string): Promise<PlaceSearchResult[]> {
  const wire = await apiRequest<PlaceSearchResponseWire>(
    `/api/v1/guides/${guideId}/place-search?q=${encodeURIComponent(query)}`
  );
  return wire.results.map((r) => ({
    label: r.label,
    latitude: r.latitude,
    longitude: r.longitude,
    placeId: r.place_id,
  }));
}

File: backend/app/schemas/place_search.py
from pydantic import BaseModel


class PlaceSearchResult(BaseModel):
    """One candidate the guide might mean, from
    app/services/places/google_provider.py. `label` is the full
    human-readable name to show in the suggestion list -- the mobile app
    never constructs its own display string from raw parts."""

    label: str
    latitude: float
    longitude: float
    # The provider's own id for this place, when picking THIS specific
    # candidate is unambiguous enough to carry through as
    # location_source='user_selected' provenance (see Submission.
    # external_place_id). None when the provider couldn't supply one.
    place_id: str | None = None


class PlaceSearchResponse(BaseModel):
    results: list[PlaceSearchResult]
